### ISSUE TITLE
Weirdness with the sticky bits (not really an issue, just documenting) -

### ISSUE THREAD

**Issue by thewacokid** [pftool/pftool#33]
Since we have pftool set with the sticky bit for various reasons, things like LD_LIBRARY_PATH get stripped when a user invokes pftool.  This is annoying, since we generally build and run using the modules system like standard clusters do.

The solution is to embed the OpenMPI dynamic library paths into the executable:
https://www.thanassis.space/tricks.html#smartdynamic

This way, users will always get the correct MPI library regardless of their environment, and regardless of whether or not their environment has been stripped due to the sticky bit security stuff.

**Comment by garygrider** [comment]
It only runs on batch nodes
Could you just preface pftool with a module of your own that sets the env how you want it.  It goes away when pftool ends anyway?

________________________________

**Comment by thewacokid** [comment]
It runs on the batch node *as* the user though, so if they load modules (really, we load modules for them) they get ignored when invoking anything with a stick bit.  Embedding the library path in the binary solves it, I just wanted to ensure we didn't forget why it doesn't work out the box. :)

**Comment by garygrider** [comment]
Confused
If they load a module on an interactive node . Then ssh to a batch node, their env comes with them?
How?

They can't run module load on the batch node.
So the only way they could get a module loaded on the batch node is .cshrc ?
We probably don't want to run .cshrc anyway, isn't that a worry?

________________________________

**Comment by thewacokid** [comment]
We can enforce default modules on the batch nodes (since they're coming in over SSH).  The problem is no matter what we do to their environment, it gets cleared out since users cannot invoke something that has the sticky bit set with their environment intact.

IE: LD_LIBRARY_PATH gets wiped whenever they invoke pftool, even if **we** fix their environment on the batch nodes to have the right paths in place.

This is not a security concern, it's a "I just built pftool and users can't run it" concern. Unless we embed the dynamic search path into the binary, there's no way for a user to run pftool.

**Comment by jti-lanl** [comment]
I think it would be pretty easy to implement what you suggest.  (Looking into it.)

I'm picturing that it would become a configuration-option, to specify what paths are built into the executable with -R.
But does that mean that we'll get complaints like "Hey, no matter how hard I load the OpenMPI module I want, pftool is still using the wrong one!"

**Comment by garygrider** [comment]
Wow
The sticky bit doesn't allow linked libs of any sort

________________________________

**Comment by garygrider** [comment]
We won't get that complaint
Pftool runs on batch nodes and only on batch node
The user can run no other mpi program there. Further they shouldn't be able to run mpi programs at all anywhere on the fta's

So I don't see how modules or anyone of that sort matter, we don't let them run anything on the batch nodes
Even if they have a script they submit when we get a scheduled, the script runs on interactive or special interactive fta's

________________________________

**Comment by jti-lanl** [comment]
Okay, but they might not know that.  I guess we have the right to tell them what MPI is used with pfcp.  You're right, not a problem.

**Comment by thewacokid** [comment]
Right, this is purely just an "it needs to work the way we want it to" kind of problem.  I'm not sure there's an elegant way besides having the search path that we want to enforce in the binary.

Short term, I did just this (embedded the search path).  I'm certainly open to better ways to do it, but this feels pretty easy and failsafe.

**Comment by jti-lanl** [comment]
Wow.  Did you cut-and-paste then modify the compile-line generated by the configured makefile (in which case we would still want to fix the configure to give you an easier method) ?

 Or, did you find a way to get configure to do this for you (in which case we wouldn't) ?

**Comment by thewacokid** [comment]
I jammed it into MARFS_LDFLAGS. Quick and dirty!

**Comment by thewacokid** [comment]
This has been fixed for a long time now. :)
